File: HW-4/week5_ANN_ex3_real_v1_nobatch.m
clear
clc
close all

% now make a real experiment

[ trainingSet, testSet, y1, y2 ] = getIrisData();

save trainingSet trainingSet
save testSet testSet
save y1 y1
save y2 y2
% train data
input = trainingSet;
% train labels
groundTruth = y1;

% Learning coefficient
coeff = 0.1;
% Number of learning iterations
iterations = 10000;

% Calculate weights randomly using seed.
rand('state',sum(100*clock));
%weights = -1 +2.*rand(3,3);

inputLength=size(input,1);
hiddenN=5; %  nodes of hidden layer1
outputN=size(groundTruth,2);
num_layers=3;
tol=0.1;

%initialize parameters
stack = cell(hiddenN+1,1); 
for k=1:num_layers-1   
    if k==1
        stack{k}.w = rand(hiddenN,inputLength);
        stack{k}.b = rand(hiddenN,1);    
    elseif k==num_layers-1
        stack{k}.w = rand(outputN,hiddenN);
        stack{k}.b = rand(outputN,1);
    else
        stack{k}.w = rand(hiddenN,hiddenN);
        stack{k}.b = rand(hiddenN,1);
    end
end
 
outputStack = cell(num_layers,1); %keep outputs
gradStack = cell(num_layers,1); %keep gradients

fig=figure; 
hold on;
% plot(1,1, 'b*');
% uiwait(fig,1)

for i = 1:iterations
     
    err=0;
    
    for j = 1:size(y1,1)       
      inputs=input(:,j);      
      outputStack{1}=inputs;
      % forward propagation
      for k=1:num_layers-1          
          outputStack{k+1} = stack{k}.w* outputStack{k}+stack{k}.b*-1;
          outputStack{k+1} = sigm(outputStack{k+1});
      end
    
      % backward propagation
      p = outputStack{end};
      
      epsilon = nnloss(groundTruth(j,:)', p, 1);
      cost = nnloss(groundTruth(j,:)', p, 0);
      err = err+cost;
      
     
      
      for k=num_layers:-1:2          
          gradStack{k-1}.epsilon = outputStack{k}.*(1-outputStack{k}).*epsilon;
          epsilon = stack{k-1}.w'*gradStack{k-1}.epsilon;         
      end

      % Add weight changes to original weights 
      % And use the new weights to repeat process.
      % delta = coeff*x*epsilon      
      for k=1:num_layers-1
          stack{k}.w = stack{k}.w + coeff*gradStack{k}.epsilon*outputStack{k}';
          stack{k}.b = stack{k}.b + coeff*(-1).*gradStack{k}.epsilon;          
      end

    end    
  
    if mod(i,100)==0
 
        hold on;
        plot(i,err, 'b*-');
        uiwait(fig,1)
    end
    
    if err<tol
        break;
    end
   
end

save stack stack

load stack stack
load testSet testSet
load y2 y2
%test the code
input = testSet;
tol=0.1;
groundTruth = y2;
out = zeros(size(y2,1),size(y2,2));
outputStack = cell(num_layers,1); 
count=0;
for j = 1:size(y2,1)
    inputs=input(:,j);
    outputStack{1}=inputs;
    % forward propagation
    for k=1:num_layers-1
      outputStack{k+1} = stack{k}.w* outputStack{k}+stack{k}.b*-1;
      outputStack{k+1} = sigm(outputStack{k+1});
    end
    out(j,:)=outputStack{end}';
    epsilon = (groundTruth(j,:) - out(j,:));
    err=sum(epsilon.^2);
    if err<tol
        count=count+1;
    end
end
   
disp('accuracy of system:')
acc = (count/size(out,1))*100
   

File: HW-4/week5_ANN_ex3_real_v1_nobatch.py
# -*- coding: utf-8 -*-
"""
Created on Sat Dec 14 11:33:36 2019

@Name-Surname: NUR SULTAN BOLEL
@Homework: Homework-4
"""

import getIrısData as getData
import numpy as np
import sigm as S
import nnloss as NL
import matplotlib.pyplot as plt

arr_x=np.zeros((1, 100))
arr_y=np.zeros((1, 100))
counter=0

trainingSet, testSet, y1, y2  = getData.getIrisData()

np.save('trainingSet', trainingSet)
np.save('testSet', testSet)
np.save('y1', y1)
np.save('y2', y2)

input = trainingSet

groundTruth = y1

coeff = 0.1
iterations = 10000

inputLength=input.shape[0]
hiddenN=5; 
outputN=groundTruth.shape[1]
num_layers=3
tol=0.1

mystack_1_w=np.random.rand(hiddenN,inputLength)
mystack_1_b=np.random.rand(hiddenN,1)

mystack_2_w=np.random.rand(outputN,hiddenN)
mystack_2_b=np.random.rand(outputN,1)

outputStack1 = np.zeros((5, 1))
outputStack2 = np.zeros((5, 1))
outputStack3 = np.zeros((3, 1))
p=np.zeros((3, 1))

gradStack1_epsilon = np.zeros((5, 1))
gradStack2_epsilon = np.zeros((3, 1))
inputs=np.zeros((5, 1))

plt.Figure()

for i in range(iterations):
    err=0
    for j in range(y1.shape[0]):
        inputs[:,0]=input[:,j]
        outputStack1 = inputs
        
        outputStack2=np.subtract(np.dot(mystack_1_w,outputStack1),mystack_1_b)
        outputStack2=S.sigm(outputStack2)
        
        outputStack3=np.subtract(np.dot(mystack_2_w,outputStack2),mystack_2_b)
        outputStack3=S.sigm(outputStack3)
        
        p = outputStack3
        epsilon = NL.nnloss(groundTruth[j:j+1,:].T, p, 1)
        cost = NL.nnloss(groundTruth[j:j+1,:].T, p, 0);
        err = err+cost

        gradStack2_epsilon=np.multiply(np.multiply(outputStack3,(1 - outputStack3)),epsilon)
        epsilon=np.dot(mystack_2_w.T,gradStack2_epsilon)
        
        gradStack1_epsilon=np.multiply(np.multiply(outputStack2,(1 - outputStack2)),epsilon)
        epsilon=np.dot(mystack_1_w.T,gradStack1_epsilon)
        
        mystack_1_w = mystack_1_w + np.multiply(coeff, np.dot(gradStack1_epsilon, (outputStack1.T)))
        mystack_1_b = mystack_1_b + np.multiply((coeff*(-1)), gradStack1_epsilon)
        
        mystack_2_w = mystack_2_w + np.multiply(coeff, np.dot(gradStack2_epsilon, (outputStack2.T)))
        mystack_2_b = mystack_2_b + np.multiply((coeff*(-1)), gradStack2_epsilon)
        
    if i%100 == 0:
        arr_x[0,counter] = i
        arr_y[0,counter] = err
        counter+=1

    if err < tol:
        break
    
plt.plot(arr_x[0,:],arr_y[0,:], 'b-')
plt.axis([0, 10000, 0, 10])
plt.show()
    
np.save('mystack_1_w', mystack_1_w)
np.save('mystack_2_w', mystack_2_w)
np.save('mystack_1_b', mystack_1_b)
np.save('mystack_2_b', mystack_2_b)

testSet = np.load('testSet.npy')
y2 = np.load('y2.npy')
#test the code
input = testSet

tol=0.1
groundTruth = y2
out=np.zeros((y2.shape[0],y2.shape[1]))
outputStack = np.random.rand(num_layers,1) 
count=0

for j in range(y2.shape[0]):

    inputs[:,0]=input[:,j]
    outputStack1 = inputs
    
    outputStack2=np.subtract(np.dot(mystack_1_w,outputStack1),mystack_1_b)
    outputStack2=S.sigm(outputStack2)
        
    outputStack3=np.subtract(np.dot(mystack_2_w,outputStack2),mystack_2_b)
    outputStack3=S.sigm(outputStack3)
    
    out[j,:]=outputStack3.T
    epsilon=np.subtract(groundTruth[j,:],out[j,:])
    err=np.sum(pow(epsilon,2))
    
    if err<tol:
        count=count+1
        
acc = (count/out.shape[0])*100    
print('accuracy of system: ', acc)